Add support for unit conversion in image writers

mraspaud opened the pull request:
This PR allows converting the units of the data when writing data to disk, in particular for geotiff floating point files.

This uses pint-xarray in the background.

 - [x] Tests added <!-- for all bug fixes or enhancements -->

djhoese commented on satpy/writers/__init__.py:
```diff
@@ -865,14 +865,17 @@ def separate_init_kwargs(cls, kwargs):
         functions for more details on the arguments passed to this method.
 
         """
+        if units is not None:
+            import pint_xarray  # noqa
+            dataset = dataset.pint.quantify().pint.to(units)
```
Not sure if you wanted this PR reviewed yet, but do you plan on adding error handling? Both for units being specified but pint not being available and for the case where the units can't be converted.

Oh and maybe `units` could be a dictionary? Mapping input units to output units? I'm thinking of the `save_datasets` case where you have visible and IR bands or any number of level 2 products.

djhoese replied:
> I haven't planned for the multiple saving case. This PR is just for save_dataset

...but...but...it would benefit my projects :cry:

mraspaud replied:
I understand, but this will require modification of the other function (`save_datasets`) and other tests. I suggest we do this in another PR.

djhoese replied:
Yeah understood. Just complaining for the sake of complaining.

mraspaud commented on satpy/writers/__init__.py:
```diff
@@ -781,6 +781,10 @@ def save_dataset(self, dataset, filename=None, fill_value=None,
                             If `False` return either a :doc:`dask:delayed`
                             object or tuple of (source, target). See the
                             return values below for more information.
+            units (srting or None): If not None, will convert the dataset to
```
```suggestion
            units (str or None): If not None, will convert the dataset to
```

gerritholl commented on satpy/writers/__init__.py:
```diff
@@ -865,14 +869,17 @@ def separate_init_kwargs(cls, kwargs):
         """
+        if units is not None:
+            import pint_xarray  # noqa
+            dataset = dataset.pint.quantify().pint.to(units)
```
```suggestion
            dataset = dataset.pint.quantify().pint.to(units).pint.dequantify()
```

this way the `units` attribute is retained.  I'm not sure if it matters for GeoTIFF, but it does for the `ninjogeotiff` writer, which inherits from the `GeoTIFF` writer, and of course it matters for the CF writer.

gerritholl replied:
Warning: unit conversion changes any integer data type to float64.  This is due to pint, not Satpy, but in Satpy this happens immediately before storing.  I wonder if there are cases where this could lead to unexpected consequences.

mraspaud replied:
Good point, I will think about this!

gerritholl commented on satpy/writers/__init__.py:
```diff
@@ -781,6 +781,10 @@ def save_dataset(self, dataset, filename=None, fill_value=None,
                             object or tuple of (source, target). See the
                             return values below for more information.
+            units (str or None): If not None, will convert the dataset to
+                                    the given unit before saving using
+                                    pint-xarray. Default is not to do any
```
```suggestion
                                    the given unit using pint-xarray before
                                    saving. Default is not to do any
```

Otherwise it sounds like it is using pint-xarray for saving.

mraspaud replied:
Indeed, thanks

gerritholl reviewed (changes requested):
I left a couple of comments and one suggested change.

I think this is good, but I'm also feeling a little bit worried about unintended consequences we're not thinking of.  Users of the ninjogeotiff writer might get confused as there will now exist both `units` and `PhysicUnit` for the writer, which sound similar but don't do the same thing.  The ninjogeotiff writer doesn't actually touch the data, but rather adapts the add_offset and scale_factor, which are writer-dependent and not standardised for all file formats.  If a user passes both, we would be correcting twice.  I wonder if we should adapt the ninjogeotiff writer to warn for this, or even completely change its unit conversion routines and delegate them ot the base writer.

mraspaud commented:
@gerritholl I addressed your comment and suggestions. Regarding the ninjogeotiff writer, it would make sense to use this new feature, right? Is there anything we need to change here to make this possible?

gerritholl commented:
The ninjogeotiff writer does not actually change values.  It just changes the offset for the special case of conversion between °C and K.  No other conversion is supported.  I suppose changing the values should lead to identical files in the end.  There was a reason why the ninjogeotiff writen (and before it the ninjotiff writer) didn't change the values; I think it was because the code in question was reached after enhancements had already been applied.

mraspaud commented:
> The ninjogeotiff writer does not actually change values. It just changes the offset for the special case of conversion between °C and K. No other conversion is supported. I suppose changing the values should lead to identical files in the end. There was a reason why the ninjogeotiff writen (and before it the ninjotiff writer) didn't change the values; I think it was because the code in question was reached after enhancements had already been applied.

That makes sense. So you are ok with merging this as it is?
